userspace/libscap/scap_savefile.c

```diff
@@ -2968,3 +2977,107 @@ void scap_fseek(scap_t *handle, uint64_t off)
 			return;
 	}
 }
+
+int scap_reader_read_capture_plugin(scap_reader_t *r, uint8_t* buf, uint32_t len)
+{
+	uint32_t nread = 0;
+	ss_plugin_rc rc;
+
+	// If seek has legitimately been used backward, we first
+	// replay the last bytes of the previous read in the buffer.
+	if (r->m_seek_buffer_ptr != NULL && r->m_seek_buffer_len > 0)
+	{
+		nread = MIN(len, r->m_seek_buffer_len);
+		memcpy(buf, r->m_seek_buffer_ptr, nread);
+		r->m_seek_buffer_len -= nread;
+		r->m_seek_buffer_ptr += nread;
+		r->m_total_bytes_read += nread;
+		buf += nread;
+		len -= nread;
+	}
+	
+	// We want to loop on the plugin_read function because the plugin
+	// could return an SS_PLUGIN_TIMEOUT rc, which does not make sense
+	// for capture plugins. In that case, we sleep for a little and then
+	// resume the loop on plugin_read(). This still ensures that the read
+	// conceretly happens only once.
+	// This also skips the plugin_read() call if len is zero at this point.

```

Comment:
Should we change the API then? Mandate that read() always returns data, potentially waiting until it does.